apm-agent-core/src/main/java/co/elastic/apm/agent/collections/WeakConcurrentSupplierImpl.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.collections;
+
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.sdk.weakconcurrent.DetachedThreadLocal;
+import co.elastic.apm.agent.sdk.weakconcurrent.WeakConcurrent;
+import co.elastic.apm.agent.sdk.weakconcurrent.WeakMap;
+import co.elastic.apm.agent.sdk.weakconcurrent.WeakSet;
+import com.blogspot.mydailyjava.weaklockfree.AbstractWeakConcurrentMap;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentSet;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * The canonical place to get a new instance of a {@link WeakMap}, {@link WeakSet}, or {@link DetachedThreadLocal}.
+ * Do not instantiate a {@link AbstractWeakConcurrentMap} directly to benefit from the global cleanup of stale entries.
+ */
+public class WeakConcurrentSupplierImpl implements WeakConcurrent.WeakConcurrentSupplier {
+
+    private static final WeakConcurrentSet<AbstractWeakConcurrentMap<?, ?, ?>> registeredMaps = new WeakConcurrentSet<>(WeakConcurrentSet.Cleaner.INLINE);
+
+    public static <K, V extends AbstractSpan<?>> WeakMap<K, V> createWeakSpanMap() {
```

Comment:
[minor] can we pull this method into parent interface and add `WeakConcurrent.createWeakSpanMap()` to avoid accesing the `WeakConcurrentSupplierImpl` class directly ?

Reply:
Unfortunately, we can't. That's because the plugin-sdk module is agnostic of the tracer implementation so it doesn't have a dependency on the core module (which is where AbstractSpan is located).